In [122]:
import pandas as pd
import numpy as np

# Загрузка и подготовка данных

In [123]:
#импортируем подготовленный датасет
pokemon_df = pd.read_csv('../data/pokemon_preprocessed.csv')

In [124]:
#убеждаемся, что загрузка прошла успешно
pokemon_df.head()

Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  \
0    318  45      49       49       65       65     45           1          0   
1    405  60      62       63       80       80     60           1          0   
2    525  80      82       83      100      100     80           1          0   
3    625  80     100      123      122      120     80           1          0   
4    309  39      52       43       60       50     65           1          0   

   Bug  ...  Ghost  Grass  Ground  Ice  Normal  Poison  Psychic  Rock  Steel  \
0    0  ...      0      1       0    0       0       1        0     0      0   
1    0  ...      0      1       0    0       0       1        0     0      0   
2    0  ...      0      1       0    0       0       1        0     0      0   
3    0  ...      0      1       0    0       0       1        0     0      0   
4    0  ...      0      0       0    0       0       0        0     0      0   

   Water  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 27 columns]

In [125]:
#нормируем вещественные признаки
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(pokemon_df.iloc[:,0:7])
scaled_df = pd.DataFrame(np.array(scaler.transform(pokemon_df.iloc[:,0:7])), columns = ["Total","HP","Attack","Defense","Sp_Atk","Sp_Def","Speed"], index = pokemon_df.index)

In [126]:
pokemon_df = pd.concat([scaled_df, pokemon_df.iloc[:,7:]], axis = "columns")
pokemon_df

Total        HP    Attack   Defense    Sp_Atk    Sp_Def     Speed  \
0   -0.976765 -0.950626 -0.924906 -0.797154 -0.239130 -0.248189 -0.801503   
1   -0.251088 -0.362822 -0.524130 -0.347917  0.219560  0.291156 -0.285015   
2    0.749845  0.420917  0.092448  0.293849  0.831146  1.010283  0.403635   
3    1.583957  0.420917  0.647369  1.577381  1.503891  1.729409  0.403635   
4   -1.051836 -1.185748 -0.832419 -0.989683 -0.392027 -0.787533 -0.112853   
..        ...       ...       ...       ...       ...       ...       ...   
795  1.375429 -0.754692  0.647369  2.443765  0.831146  2.808099 -0.629341   
796  2.209541 -0.754692  2.497104  1.160233  2.665905  1.369846  1.436611   
797  1.375429  0.420917  0.955658 -0.444182  2.360112  2.088973  0.059310   
798  2.042718  0.420917  2.497104 -0.444182  2.971699  2.088973  0.403635   
799  1.375429  0.420917  0.955658  1.481116  1.748526  0.650720  0.059310   

     Generation  Legendary  Bug  ...  Ghost  Grass  Ground  Ice  Normal  \
0             1          0    0  ...      0      1       0    0       0   
1             1          0    0  ...      0      1       0    0       0   
2             1          0    0  ...      0      1       0    0       0   
3             1          0    0  ...      0      1       0    0       0   
4             1          0    0  ...      0      0       0    0       0   
..          ...        ...  ...  ...    ...    ...     ...  ...     ...   
795           6          1    0  ...      0      0       0    0       0   
796           6          1    0  ...      0      0       0    0       0   
797           6          1    0  ...      1      0       0    0       0   
798           6          1    0  ...      0      0       0    0       0   
799           6          1    0  ...      0      0       0    0       0   

     Poison  Psychic  Rock  Steel  Water  
0         1        0     0      0      0  
1         1        0     0      0      0  
2         1        0     0      0      0  
3         1        0     0      0      0  
4         0        0     0      0      0  
..      ...      ...   ...    ...    ...  
795       0        0     1      0      0  
796       0        0     1      0      0  
797       0        1     0      0      0  
798       0        1     0      0      0  
799       0        0     0      0      1  

[800 rows x 27 columns]

In [127]:
#создадим датафреймы X и y
y = pokemon_df.Legendary
pokemon_df = pokemon_df.drop(['Legendary'], axis=1)
X = pokemon_df

In [128]:
from sklearn.model_selection import train_test_split
#разбиваем всю выборку на обучающую и тестовую
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 2)

# Метод k-ближайших соседей (K-Nearest Neighbors)

In [129]:
from sklearn.neighbors import KNeighborsClassifier

clf = KNeighborsClassifier(n_neighbors=5) # создаем классификатор
clf.fit(X_train, y_train) # обучаем классификатор

KNeighborsClassifier()

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

y_pred = clf.predict(X_test) # предсказываем ответы с помощью классификатора
print(confusion_matrix(y_test, y_pred)) 
print(classification_report(y_test, y_pred))

Последующие классификаторы создаются, обучаются и предсказывают ответ аналогично данному.

# Классификатор дерева решений (Decision Tree Classifier)

In [131]:
from sklearn import tree

clf = tree.DecisionTreeClassifier()
clf = clf.fit(X_train, y_train)

In [132]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[144   5]
 [  0  11]]
              precision    recall  f1-score   support

           0       1.00      0.97      0.98       149
           1       0.69      1.00      0.81        11

    accuracy                           0.97       160
   macro avg       0.84      0.98      0.90       160
weighted avg       0.98      0.97      0.97       160



In [133]:
#decision tree CART model

clf = tree.DecisionTreeClassifier(criterion = "gini", random_state = 2)
clf = clf.fit(X_train, y_train)

In [134]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[144   5]
 [  0  11]]
              precision    recall  f1-score   support

           0       1.00      0.97      0.98       149
           1       0.69      1.00      0.81        11

    accuracy                           0.97       160
   macro avg       0.84      0.98      0.90       160
weighted avg       0.98      0.97      0.97       160



In [135]:
#decision tree C50 model

clf = tree.DecisionTreeClassifier(criterion="entropy", random_state = 2)
clf = clf.fit(X_train, y_train)

In [136]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[144   5]
 [  1  10]]
              precision    recall  f1-score   support

           0       0.99      0.97      0.98       149
           1       0.67      0.91      0.77        11

    accuracy                           0.96       160
   macro avg       0.83      0.94      0.87       160
weighted avg       0.97      0.96      0.97       160



# Наивный байесовский классификатор (Naive Bayes)

In [137]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB

clf = GaussianNB()
clf.fit(X_train, y_train)

GaussianNB()

In [138]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[82 67]
 [ 0 11]]
              precision    recall  f1-score   support

           0       1.00      0.55      0.71       149
           1       0.14      1.00      0.25        11

    accuracy                           0.58       160
   macro avg       0.57      0.78      0.48       160
weighted avg       0.94      0.58      0.68       160



In [139]:
clf = BernoulliNB()
clf.fit(X_train, y_train)

BernoulliNB()

In [140]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[139  10]
 [  1  10]]
              precision    recall  f1-score   support

           0       0.99      0.93      0.96       149
           1       0.50      0.91      0.65        11

    accuracy                           0.93       160
   macro avg       0.75      0.92      0.80       160
weighted avg       0.96      0.93      0.94       160



# Метод опорных векторов (Support Vector Machines)

In [141]:
from sklearn import svm
clf = svm.SVC()
clf.fit(X_train, y_train)

SVC()

In [142]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[145   4]
 [  3   8]]
              precision    recall  f1-score   support

           0       0.98      0.97      0.98       149
           1       0.67      0.73      0.70        11

    accuracy                           0.96       160
   macro avg       0.82      0.85      0.84       160
weighted avg       0.96      0.96      0.96       160



# Логистическая регрессия (Logistic Regression)

In [143]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression()
clf.fit(X_train, y_train)

LogisticRegression()

In [144]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[146   3]
 [  3   8]]
              precision    recall  f1-score   support

           0       0.98      0.98      0.98       149
           1       0.73      0.73      0.73        11

    accuracy                           0.96       160
   macro avg       0.85      0.85      0.85       160
weighted avg       0.96      0.96      0.96       160



# Случайный лес (Random Forest)

In [145]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier()
clf.fit(X_train, y_train)

RandomForestClassifier()

In [146]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[146   3]
 [  0  11]]
              precision    recall  f1-score   support

           0       1.00      0.98      0.99       149
           1       0.79      1.00      0.88        11

    accuracy                           0.98       160
   macro avg       0.89      0.99      0.93       160
weighted avg       0.99      0.98      0.98       160



# Градиентный бустинг (Gradient Boosting) 

In [147]:
from sklearn.ensemble import GradientBoostingClassifier

clf = GradientBoostingClassifier()
clf.fit(X_train, y_train)

GradientBoostingClassifier()

In [148]:
y_pred = clf.predict(X_test)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[145   4]
 [  0  11]]
              precision    recall  f1-score   support

           0       1.00      0.97      0.99       149
           1       0.73      1.00      0.85        11

    accuracy                           0.97       160
   macro avg       0.87      0.99      0.92       160
weighted avg       0.98      0.97      0.98       160



# Подбор гиперпараметров

In [149]:
from sklearn.model_selection import GridSearchCV

parameters = {'n_neighbors':[1, 2, 3, 4, 5]}

knn = KNeighborsClassifier()

clf = GridSearchCV(knn, parameters)

clf.fit(X_train, y_train)

clf.best_params_

{'n_neighbors': 5}

В итоге оказалось, что мы изначально выбрали наилучший гиперпараметр (количество соседей равное пяти).

# Итоги

Подводя итоги, можно сказать, что самым лучшим классификатором в данном случае оказался случайный лес со значением метрики accuracy, равным 0.98, то есть классификатор ошибается лишь в 2 процентах случаев (3 покемона из 160), это более чем достойный результат.
В то же время можем сказать, что худшим оказался наивный байесовский классификатор Гаусса, который ошибается в 42 процентах случаев, что неприемлимо. Остальные модели оказались примерно такими же качественными, как и случайный лес, но со значением метрики accuracy немного меньше, 0.96-0.97 (за исключением наивного байесовского классификатора Бернулли, accuracy которого оказалась равна 0.93).